### PR TITLE
CSP: Parse details in host-scheme

### PR DESCRIPTION
**Pull request by morlovich**
This is a bit complicated since those aren't URLs proper, as they can have wildcards in spot and 
can drop the scheme.

### CODE REVIEW

File: net/instaweb/rewriter/csp.cc [hunk 1]
```diff
@@ -49,43 +49,139 @@ inline bool IsAsciiAlpha(char ch) {
...
+  // Either way, it's not a valid scheme-source at this point, even if it's
+  // a valid host-source
+  return false;
+}
+
 CspSourceExpression CspSourceExpression::Parse(StringPiece input) {
   TrimCspWhitespace(&input);
   if (input.empty()) {
-    return CspSourceExpression(kUnknown);
+    return CspSourceExpression();
```
Comment (oschaaf) [inline net/instaweb/rewriter/csp.cc:101]: I wonder, will we need to be able to discern between unknown and empty/whitespace ?

File: net/instaweb/rewriter/csp_test.cc [hunk 2]
```diff
@@ -65,31 +85,119 @@ TEST(CspParseSourceTest, NonQuoted) {
...
+      CspSourceExpression(
+          CspSourceExpression::kHostSource,
+          CspSourceExpression::UrlData("https", "*", "443", "/foo.js")),
+      CspSourceExpression::Parse("https://*:443/foo.js"));
+
+  EXPECT_EQ(
+      CspSourceExpression(CspSourceExpression::kUnknown),
+      CspSourceExpression::Parse("https://*:443?foo.js"));
+}
+
```
Comment (oschaaf) [inline net/instaweb/rewriter/csp_test.cc:196]: Maybe also test our expectations for `CspSourceExpression::ParseQuoted("\"\"")` ?

File: net/instaweb/rewriter/csp.cc (hunk 1 above)
Comment (morlovich) [reply net/instaweb/rewriter/csp.cc:101]: Nope, basically either one won't authorize anything.

File: net/instaweb/rewriter/csp_test.cc (hunk 2 above)
Comment (morlovich) [reply net/instaweb/rewriter/csp_test.cc:196]: Double quotes don't mean anything in this syntax, so any particular reason this would be interesting?

File: net/instaweb/rewriter/csp.cc (hunk 1 above)
Comment (oschaaf) [reply net/instaweb/rewriter/csp.cc:101]: I see, thanks

File: net/instaweb/rewriter/csp_test.cc (hunk 2 above)
Comment (oschaaf) [reply net/instaweb/rewriter/csp_test.cc:196]: `CspSourceExpression::ParseQuoted("\"\"")` -> should have been `CspSourceExpression::ParseQuoted("''")`
But either way, I noticed there's already a test that inputs only whitespace, and this one wouldn't add value on top of that, so please ignore

Comment (morlovich) [reply net/instaweb/rewriter/csp_test.cc:196]: Oh, good idea, this is an edge case, so added.

**Comment by morlovich** [comment]
Note: there is a more extensive/later version in https://github.com/pagespeed/mod_pagespeed/tree/morlovich-csp-urls3, it might make sense to review that instead.

File: net/instaweb/rewriter/csp.cc [hunk 3]
```diff
@@ -49,43 +49,139 @@ inline bool IsAsciiAlpha(char ch) {
           ((ch >= 'A') && (ch <= 'Z')));
 }
 
+inline bool IsAsciiDigit(char ch) {
+  return ((ch >= '0') && (ch <= '9'));
+}
```
Comment (jmarantz) [inline net/instaweb/rewriter/csp.cc]: put this in string_util.h next to IsHexDigit?  Rename to IsDecimalDigit?

**Review by jmarantz — COMMENTED** [review]
qq: have you valgrind'd & msan'd your tests for this?  I don't see anything suspicious but it seems worthwhile.

also maybe fuzzing?

still reading...

**Comment by morlovich** [comment]
Hmm, I've done fuzzing, but probably not in the latest revision ---
basically by copying the files to the Google build system and using the
support there.
Can also copy them to the Chrome build system and do the same :). Probably
should redo for latest.

File: net/instaweb/rewriter/csp.cc [hunk 4]
```diff
@@ -49,43 +49,139 @@ inline bool IsAsciiAlpha(char ch) {
...
+  if (!IsAsciiAlpha((*input)[0])) {
+    return false;
+  }
+
+  size_t pos = 1;
+  while (pos < input->size() &&
+         (IsAsciiAlphaNumeric((*input)[pos])
+          || (*input)[pos] == '+'
+          || (*input)[pos] == '-'
+          || (*input)[pos] == '.')) {
```
Comment (jmarantz) [inline net/instaweb/rewriter/csp.cc]: it annoys me slightly that (*input)[pos] is repeated 4x :)

Maybe just factor out the per-char check to an inline helper?

File: net/instaweb/rewriter/csp.cc [hunk 5]
```diff
@@ -49,43 +49,139 @@ inline bool IsAsciiAlpha(char ch) {
...
+    }
+
+    if (host_part.starts_with("*")) {
+      if (!host_part.starts_with("*.") && (host_part != "*")) {
+        return CspSourceExpression();
+      }
+    }
+
+    // Start on port-part, if any
+    if (input.starts_with(":")) {
```
Comment (jmarantz) [inline net/instaweb/rewriter/csp.cc]: input[0] == ':'

File: net/instaweb/rewriter/csp.cc [hunk 6]
```diff
@@ -49,43 +49,139 @@ inline bool IsAsciiAlpha(char ch) {
...
+    }
+
+    // Verify accumulated host-part is valid.
+    StringPiece host_part(result.url_data().host_part);
+    if (host_part.empty()) {
+      return CspSourceExpression();
+    }
+
+    if (host_part.starts_with("*")) {
+      if (!host_part.starts_with("*.") && (host_part != "*")) {
```
Comment (jmarantz) [inline net/instaweb/rewriter/csp.cc]: if (host_part[0] == '*' && host_part.size() > 1 && host_part[1] != '.' )

File: net/instaweb/rewriter/csp.cc (hunk 5 above)
Comment (morlovich) [reply net/instaweb/rewriter/csp.cc]: input could be empty here.

**Review by jmarantz — APPROVED** [review]
It'd be great to get some coverage tests on this for your unit tests.

File: net/instaweb/rewriter/csp.cc (hunk 3 above)
Comment (morlovich) [reply net/instaweb/rewriter/csp.cc]: Done

File: net/instaweb/rewriter/csp.cc (hunk 4 above)
Comment (morlovich) [reply net/instaweb/rewriter/csp.cc]: Donr

File: net/instaweb/rewriter/csp.cc (hunk 6 above)
Comment (morlovich) [reply net/instaweb/rewriter/csp.cc]: Thanks, this is much better.
